### PR TITLE
Milestone-3

### PR DESCRIPTION
**Pull request by David405**
Created the game Logic

### CODE REVIEW

File: lib/players.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class Player
+  require_relative 'board'
```
Comment (maelfosso) [inline lib/players.rb]: Put it at the beginning of the file, please.

File: lib/players.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  def get_move(board)
+    puts "#{name}, choose a position to place your #{marker}"
+    position = gets.chomp
+
+    until position =~ /^[0-8]{1}$/ && (board.board[position.to_i] != 'x' && board.board[position.to_i] != 'o')
+      puts 'Enter a number that is valid'
+      position = gets.chomp
+    end
+  end
+end
```
Comment (maelfosso) [inline lib/players.rb]: Please, move those two methods into the `Board` class.

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  private
+
+  def display_board(board)
+    puts ''
+    puts " #{board[0]} | #{board[1]} | #{board[2]} "
+    puts '--- --- ---'
+    puts " #{board[3]} | #{board[4]} | #{board[5]} "
+    puts '--- --- ---'
+    puts " #{board[6]} | #{board[7]} | #{board[8]} "
+  end
```
Comment (maelfosso) [inline lib/board.rb]: Move that function into `bin/main.rb`
You will use it 
```ruby
def display_board(board)
   ....
end

display_board(game.board)
```

File: bin/main.rb [hunk 4]
```diff
@@ -1,79 +1,94 @@
-#!/usr/bin/env ruby
-=begin
-class Board 
-    attr_reader :board
+class Game 
+    load 'lib/board.rb'
+    load 'lib/players.rb'
```
Comment (maelfosso) [inline bin/main.rb]: Use `require_relative` and at the beginning of the file

File: bin/main.rb [hunk 5]
```diff
@@ -1,79 +1,94 @@
...
-=end
+        puts 'Player 2, enter name:'
+        name2 = gets.chomp
+        @player2 = Player.new(name2, 'o')
 
-#Print board with registered move
-p "display_board"
+        puts '#{@player1.name.capitalize!}, you are x!'
+        puts '#{@player2,name.capitalize!}, you are o!'
+    end
```
Comment (maelfosso) [inline bin/main.rb]: That function should stay into `bin/main`. No need to put it into a class

File: bin/main.rb [hunk 6]
```diff
@@ -1,79 +1,94 @@
...
 
-#The above steps would be repeated
-=begin
-while !WIN_COMBINATION do
-    "step 1" & "step 2"
-end
-=end
+    def create_board
+        @board = Board.new
+    end
```
Comment (maelfosso) [inline bin/main.rb]: Move it into `Game.initialize`

File: bin/main.rb [hunk 7]
```diff
@@ -1,79 +1,94 @@
...
+        @player1 = nil
+        @player2 = nil
+        @board = nil
+        @game_over = false
+        @turn = 0
+
+        setup_players
+        create_board
+        play_game
     end
```
Comment (maelfosso) [inline bin/main.rb]: Change it like. 
```ruby
def initialize(name1, name2)
    @player1 = Player.new(name1)
    @player2 = Player.new(name2)
    @board = Board.new

    .....
end
```

File: bin/main.rb (hunk 5 above)
Comment (maelfosso) [reply bin/main.rb]: But you should initialize the players into `Game.initialize`

File: bin/main.rb [hunk 8]
```diff
@@ -1,79 +1,94 @@
...
-#Print board with registered move
-p "display_board"
+    def play_game
+        until @game_over
+            @player2.move(@board)
+            check_status(@player2)
+            @player1.move(@board)
+            check_status(@player1)
+        end
+    end
```
Comment (maelfosso) [inline bin/main.rb]: Very good function :+1: 

Don't you think that it's more  preferable to say 
```ruby
@game.move(input, @player2)
```
With `game.move` calling `board.move` ?

File: bin/main.rb [hunk 9]
```diff
@@ -1,79 +1,94 @@
...
+            [0, 3, 6], [1, 4, 7], [2, 5, 8],
+            #diagonal
+            [0, 4, 8], [2, 4, 6]
+        ]
+
+        player_won = win_combinations.any? do |c|
+            c.all? do |i|
+                @board.board[i] == player.marker
+            end
+        end
```
Comment (maelfosso) [inline bin/main.rb]: You should move that into `Board` class

File: bin/main.rb [hunk 10]
```diff
@@ -1,79 +1,94 @@
...
-p "display_board"
+    def end_game(player)
+        if player == 'tie'
+            puts "It's a tie."
+        else
+            puts '#{player.name} won!'
+        end
+        @game_over = true
+        reset_game
+    end
```
Comment (maelfosso) [inline bin/main.rb]: That snippet of code should be a method into the `Game` class

File: lib/players.rb (hunk 2 above)
Comment (David405) [reply lib/players.rb]: done

File: bin/main.rb (hunk 6 above)
Comment (David405) [reply bin/main.rb]: done

File: bin/main.rb (hunk 5 above)
Comment (David405) [reply bin/main.rb]: done

File: bin/main.rb (hunk 7 above)
Comment (David405) [reply bin/main.rb]: done

File: bin/main.rb (hunk 8 above)
Comment (David405) [reply bin/main.rb]: Yeah but won't that require creating a new game method?

File: lib/players.rb (hunk 2 above)
Comment (David405) [reply lib/players.rb]: move and get_move method does not work well inside of the Board class

File: bin/main.rb [hunk 11]
```diff
@@ -1,79 +1,82 @@
...
-p "Move completed"
+repeat_game = true
 
-#move check
-=begin 
-if player_input is not found in board
-    p "invalid move"
-=end
+while repeat_game
+  new_board = Board.new if gets.chomp == "ok" || "OK"
```
[the remark this answers is not in the document]
Comment (David405) [reply bin/main.rb]: fixed

File: bin/main.rb [hunk 12]
```diff
@@ -1,79 +1,82 @@
...
+    puts "Restart game? (Enter yes or no)"
+    case gets.chomp
+    when "yes"
+      puts "Enter OK to continue."
+    when "no"
+      repeat_game = false
+      abort "Game over"
+    else
+      puts "Kindly enter yes or no"
+    end
```
[the remark this answers is not in the document]
Comment (David405) [reply bin/main.rb]: fixed

**Pull request merged**